File: backend/src/services/SocksAccountService.ts
import { getDatabase } from '../database/connection';
import { SocksAccount } from '../types';
import { generateId } from '../utils';

export class SocksAccountService {
  async getAll(): Promise<SocksAccount[]> {
    const db = getDatabase();
    const rows = await db.all('SELECT * FROM socks_accounts ORDER BY createdAt DESC');
    return rows.map(r => ({ ...r, enabled: Boolean(r.enabled) }));
  }

  async getByUsername(username: string): Promise<SocksAccount | null> {
    const db = getDatabase();
    const row = await db.get('SELECT * FROM socks_accounts WHERE username = ? AND enabled = 1', username);
    if (!row) return null;
    return { ...row, enabled: Boolean(row.enabled) };
  }

  async create(username: string, password: string, mode: 'rotate' | 'sticky', maxDelay?: number, countryFilter?: string, countryFilterMode?: 'include' | 'exclude'): Promise<SocksAccount> {
    const db = getDatabase();
    const existing = await db.get('SELECT id FROM socks_accounts WHERE username = ?', username);
    if (existing) throw new Error('用户名已存在');

    const id = generateId();
    const createdAt = new Date().toISOString();
    await db.run(
      'INSERT INTO socks_accounts (id, username, password, mode, enabled, maxDelay, countryFilter, countryFilterMode, createdAt) VALUES (?, ?, ?, ?, 1, ?, ?, ?, ?)',
      [id, username, password, mode, maxDelay ?? null, countryFilter ?? null, countryFilterMode ?? 'include', createdAt]
    );
    return { id, username, password, mode, enabled: true, maxDelay, countryFilter, countryFilterMode: countryFilterMode ?? 'include', createdAt };
  }

  async update(id: string, updates: Partial<Pick<SocksAccount, 'password' | 'mode' | 'enabled' | 'maxDelay' | 'countryFilter' | 'countryFilterMode'>>): Promise<void> {
    const db = getDatabase();
    const fields: string[] = [];
    const values: any[] = [];
    if (updates.password !== undefined) { fields.push('password = ?'); values.push(updates.password); }
    if (updates.mode !== undefined) { fields.push('mode = ?'); values.push(updates.mode); }
    if (updates.enabled !== undefined) { fields.push('enabled = ?'); values.push(updates.enabled ? 1 : 0); }
    if ('maxDelay' in updates) { fields.push('maxDelay = ?'); values.push(updates.maxDelay ?? null); }
    if ('countryFilter' in updates) { fields.push('countryFilter = ?'); values.push(updates.countryFilter ?? null); }
    if ('countryFilterMode' in updates) { fields.push('countryFilterMode = ?'); values.push(updates.countryFilterMode ?? 'include'); }
    if (fields.length === 0) return;
    values.push(id);
    await db.run(`UPDATE socks_accounts SET ${fields.join(', ')} WHERE id = ?`, values);
  }

  async delete(id: string): Promise<void> {
    const db = getDatabase();
    await db.run('DELETE FROM socks_accounts WHERE id = ?', id);
  }
}

File: backend/services/socks_account_service.py
from database import get_connection
from utils import generate_id
from datetime import datetime
from typing import List, Dict, Any, Optional


class SocksAccountService:
    """SOCKS 账号服务"""
    
    async def get_all(self) -> List[Dict[str, Any]]:
        """获取所有账号"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('SELECT * FROM socks_accounts ORDER BY createdAt DESC')
            rows = cursor.fetchall()
            return [dict(row) for row in rows]
    
    async def get_by_username(self, username: str) -> Optional[Dict[str, Any]]:
        """根据用户名获取账号"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                'SELECT * FROM socks_accounts WHERE username = ? AND enabled = 1',
                (username,)
            )
            row = cursor.fetchone()
            if not row:
                return None
            return dict(row)
    
    async def create(self, username: str, password: str, mode: str = 'rotate',
                    max_delay: Optional[int] = None, country_filter: Optional[str] = None,
                    country_filter_mode: str = 'include') -> Dict[str, Any]:
        """创建账号"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('SELECT id FROM socks_accounts WHERE username = ?', (username,))
            if cursor.fetchone():
                raise ValueError('用户名已存在')
            
            account_id = generate_id()
            created_at = datetime.now().isoformat()
            
            cursor.execute('''
                INSERT INTO socks_accounts (id, username, password, mode, enabled, maxDelay,
                                           countryFilter, countryFilterMode, createdAt)
                VALUES (?, ?, ?, ?, 1, ?, ?, ?, ?)
            ''', (account_id, username, password, mode, max_delay, country_filter,
                  country_filter_mode, created_at))
            
            return {
                'id': account_id,
                'username': username,
                'password': password,
                'mode': mode,
                'enabled': True,
                'maxDelay': max_delay,
                'countryFilter': country_filter,
                'countryFilterMode': country_filter_mode,
                'createdAt': created_at
            }
    
    async def update(self, account_id: str, updates: Dict[str, Any]) -> None:
        """更新账号"""
        fields = []
        values = []
        
        if 'password' in updates:
            fields.append('password = ?')
            values.append(updates['password'])
        
        if 'mode' in updates:
            if updates['mode'] not in ['rotate', 'sticky']:
                raise ValueError('模式无效')
            fields.append('mode = ?')
            values.append(updates['mode'])
        
        if 'enabled' in updates:
            fields.append('enabled = ?')
            values.append(1 if updates['enabled'] else 0)
        
        if 'maxDelay' in updates:
            fields.append('maxDelay = ?')
            values.append(updates['maxDelay'] if updates['maxDelay'] is not None else None)
        
        if 'countryFilter' in updates:
            fields.append('countryFilter = ?')
            values.append(updates['countryFilter'] if updates['countryFilter'] else None)
        
        if 'countryFilterMode' in updates:
            mode = updates['countryFilterMode'] if updates['countryFilterMode'] in ['include', 'exclude'] else 'include'
            fields.append('countryFilterMode = ?')
            values.append(mode)
        
        if not fields:
            return
        
        values.append(account_id)
        
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(f'UPDATE socks_accounts SET {", ".join(fields)} WHERE id = ?', values)
    
    async def delete(self, account_id: str) -> None:
        """删除账号"""
        with get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute('DELETE FROM socks_accounts WHERE id = ?', (account_id,))
